### PR TITLE
varianter: Make it properly pluggable [v2]

### PR DESCRIPTION
**Pull request by ldoktor**
This PR is inspired by https://github.com/avocado-framework/avocado/pull/1591 and is the next stage of https://github.com/avocado-framework/avocado/pull/1685 . First it adds necessary interfaces to the old `varianter`, then it makes the `varianter` pluggable in a dirty way and then it turns it into a proper plugin defined in `avocado.core.plugin_interfaces`. Along the way there are several fixes and enhancements and at the end there is the promised rewrite of the multiplexer documentation, which took me several days to put together and I hope it's understandable and clear enough.

v1: https://github.com/avocado-framework/avocado/pull/1716

Changes:

```yaml
v2: Just rebased to see whether the travis failures were false-negatives
```

### CODE REVIEW

File: avocado/core/plugin_interfaces.py [hunk 1]
```diff
@@ -213,7 +213,14 @@ class VarianterPlugin(Plugin):
     @abc.abstractmethod
     def __iter__(self):
         """
-        Yield variant in form of (variant_id, (variant, mux_path))
+        Yields all variants
```
Comment (clebergnu) [inline avocado/core/plugin_interfaces.py]: As this was introduced by a commit in the same PR, docstring such as this one should be added alongside the original commit.  In this specific case, it should then have been updated in the previous commit, when the format of the yelded variant changes to a dict.

File: avocado/core/plugin_interfaces.py [hunk 2]
```diff
@@ -235,8 +242,11 @@ def to_str(self, summary, variants, extra):
         """
         Return human readable representation
 
-        :param summary: How verbose summary to output
-        :param variants: How verbose list of variants to output
+        The summary/variants accepts verbosity where 0 means do not display
```
Comment (clebergnu) [inline avocado/core/plugin_interfaces.py]: Docstring should be squashed with the original version of this method.

File: docs/source/index.rst [hunk 3]
```diff
@@ -15,7 +15,7 @@ Contents:
    Loaders
    LoggingSystem
    Sysinfo
-   Mux
+   TestParameters
```
Comment (clebergnu) [inline docs/source/index.rst]: I have to say that the documentation changes make sense, but since this area has changed to much recently, it needs a more thorough review.  Please don't consider this brief doc review and endorsement of the overall status of the docs.

Since there are many earlier comments you can address, I'll try to do a proper doc review after this initial round.

File: avocado/core/varianter.py [hunk 4]
```diff
@@ -459,6 +459,57 @@ def data_merge(self, tree):     # pylint: disable=E0202
             return
         self.data.merge(tree)
 
+    def str_variants(self):
```
Comment (ldoktor) [inline avocado/core/varianter.py]: As I mentioned in previous PRs this code is rewritten almost from scratch. That means there are way too many changes. I did update the existing unittests, but I plan to write many more when the API settles. I asked you to be lenient towards the lack of unittest/documentation updates because I know what else is coming and we are almost there. This series actually includes the major rework of the documentation (which I kept promising in previous stages) and when we agree on this approach (I still haven't got ack on the base idea...) I can write the unit and functional tests.

Although the main functionality is already covered so I need to extend just the corner cases...

File: avocado/plugins/multiplex.py [hunk 5]
```diff
@@ -82,29 +82,12 @@ def run(self, args):
...
-                    continue
-                fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
-                for record in sorted(env):
-                    log.debug(fmt, *record)
+        if args.mux_debug:
+            # In this version `avocado_variants.debug` is not set properly,
+            # let's force-enable it before calling str_variants_long to
+            # get the expected results.
+            args.avocado_variants.debug = True
+        for line in args.avocado_variants.str_variants_long(True).splitlines():
```
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: Good catch

File: avocado/core/mux.py [hunk 6]
```diff
@@ -29,6 +36,152 @@
...
+    Object representing part of the tree from the root to leaves or another
+    multiplex domain. Recursively it creates multiplexed variants of the full
+    tree.
+    """
+
+    def __init__(self, root):
+        """
+        :param root: Root of this tree slice
+        """
+        self.pools = []
```
Comment (ldoktor) [inline avocado/core/mux.py]: Well the change is atomic, because the information about `root` is now stored in the `MuxPlugin` rather than in each `MuxTree`. I mean it would be possible to first move it here and then introduce the `MuxPlugin`, but with so many changes I thought this little change is understandable. Anyway if you insist I can rework that (causing yet another delay).

File: avocado/core/mux.py [hunk 7]
```diff
@@ -29,6 +36,152 @@
...
+        pools = itertools.product(*pools)
+        while True:
+            # TODO: Implement 2nd level filters here
+            # TODO: This part takes most of the time, optimize it
+            yield list(itertools.chain(*pools.next()))
+
+
+class MuxPlugin(object):
+    """
+    Follows the Multiplexer API to produce variants
```
Comment (ldoktor) [inline avocado/core/mux.py]: Multiplexer API is something abstract at this point and definitely nothing people should be using. Anyway one of the requirements was that the multiplexer idea should be part of the core so people could inspire and based their plugins on it. That's why I created this more-less generic class which is then used by the actual yaml_to_mux plugin.

Anyway to your question, the current API is the `__init__` where you hand over the `root` and other useful values and this class will implement the rest of the necessary things in order to produce complete varianter plugin.

File: avocado/core/parser.py [hunk 8]
```diff
@@ -145,3 +142,7 @@ def finish(self):
                     sub.choices[self.args.subcommand].error(msg)
 
             self.application.error(msg)
+
+        # Allow overriding default params by plugins
```
Comment (ldoktor) [inline avocado/core/parser.py]: Actually I can't tell anymore. I know I changed it for a reason, but maybe that reason disappeared with all the changes I made afterwards. I'll remove this (unless it breaks something in later commits, then I'll move it to that commit).

File: avocado/core/varianter.py [hunk 9]
```diff
@@ -447,132 +381,78 @@ def _skip_new_data_check(self, function, args):
...
             return
         if path is None:
             path = "/"
         if name not in self.default_params:
-            self.default_params[name] = tree.TreeNode()
+            self.default_params[name] = self.node_class()
         self.default_params[name].get_node(path, True).value[key] = value
 
-    def data_merge(self, tree):     # pylint: disable=E0202
+    def add_variants_plugin(self, plugin):     # pylint: disable=E0202
```
Comment (ldoktor) [inline avocado/core/varianter.py]: Well this function is replaced with dispatcher's entry point. So at this point I think it is quite important... I mean yes, I can just forget this step existed and immediatelly start with dispatcher-based plugin initialization, but that would be even more complex patch. Also note that this patch existed months ago before I started working on making it properly supported as a plugin interface. That's the whole problem of this PR, I change things because they were not reviewed and merged until I developed another stage which means there is 3+ moths old code combined with new requirements. In a pure agile world, all of those changes would already been applied. In the old days this would only existed in my tree and I'd send a huge update which would remove the previous and add the new iterface (which would not be possible to review, that's why people started to use patch series and due to their length they started to use agile...)

Anyway with couple of hours of (in my eyes) wasted time I can rework this to only include the `dispatcher`-based version. But I'd really prefer using that time to finish the variant-id-based-names, internal filters and selftests instead.

File: avocado/core/mux.py [hunk 10]
```diff
@@ -109,26 +109,21 @@ def update_defaults(self, defaults):
         combination.merge(self.root)
         self.variants = MuxTree(combination)
 
-    def str_variants(self):
+    def str_variants(self, summary=False, contents=False):
```
Comment (ldoktor) [inline avocado/core/mux.py]: Yep, this patch was intended for squash, if accepted (the problem is again that this PR contains several stages, because previous stages were not reviewed until I finished those new parts).

Comment (ldoktor) [inline avocado/core/mux.py]: Btw you keep pointing to some unittests you wrote but I don't see them anywhere... As I mentioned earlier I intend to extend the unittests, but at this point I didn't even know whether this approach would be acceptable or not.

File: avocado/plugins/multiplex.py [hunk 11]
```diff
@@ -35,23 +34,28 @@ def __init__(self, *args, **kwargs):
 
     def configure(self, parser):
         parser = super(Multiplex, self).configure(parser)
+        parser.add_argument("--summary", type=int, help="Verbosity of "
```
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: Well I thought about enum (short, medium, long, extensive) but with multiple plugins we don't know what's the limit and what's the usage (eg. for yaml it's actually a bitmask of different informations to show). How about making the argument optional, by default use `1` and also accept `full` (which would propagate as `99`) or integer. That should scale to custom plugins as well as it should be easy to use.

File: avocado/core/mux.py [hunk 12]
```diff
@@ -109,45 +109,54 @@ def update_defaults(self, defaults):
...
         self.variants = MuxTree(combination)
 
-    def str_variants(self, summary=False, contents=False):
+    def str_variants(self, summary=0, variants=0, use_utf8=False):
         """
-        Return human readable variants
+        Return human readable representation
+
+        :param summary: How verbose summary to output
+        :param variants: How verbose list of variants to output
```
Comment (ldoktor) [inline avocado/core/mux.py]: ack

**Comment by clebergnu** [comment]
The info on the unittests is here:

https://github.com/avocado-framework/avocado/pull/1763#pullrequestreview-20788834

2017-02-10 5:23 GMT-05:00 Lukáš Doktor <notifications@github.com>:

File: avocado/core/job.py [hunk 13]
```diff
@@ -371,7 +371,7 @@ def _log_avocado_datadir():
         job_log.info('')
 
     def _log_variants(self, variants):
-        lines = variants.str_variants(summary=1, variants=1)
+        lines = variants.to_str(summary=1, variants=1)
```
Comment (ldoktor) [inline avocado/core/job.py]: As explained earlier, it was meant for squashing, only I was not sure you'll like the name and more importantly whether you'll like the overall approach...

File: avocado/core/varianter.py [hunk 14]
```diff
@@ -377,7 +377,7 @@ def _skip_new_data_check(self, function, args):
         """
         if self.is_parsed():
             if self.ignore_new_data:
-                return
+                return True
```
Comment (ldoktor) [inline avocado/core/varianter.py]: Yep, I'll add it here.

File: avocado/core/mux.py [hunk 15]
```diff
@@ -88,20 +88,41 @@ def __iter__(self):
 
 class MuxPlugin(object):
     """
-    Follows the Multiplexer API to produce variants
+    Base implementation of Mux-like Varianter plugin and should be instantiated
+    as :class:`avocado.core.plugin_interfaces.VarianterPlugin`.
```
Comment (ldoktor) [inline avocado/core/mux.py]: sure, thanks for the English version

File: avocado/core/mux.py [hunk 16]
```diff
@@ -132,7 +132,7 @@ def update_defaults(self, defaults):
         combination.merge(self.root)
         self.variants = MuxTree(combination)
 
-    def to_str(self, summary=0, variants=0, use_utf8=False):
+    def to_str(self, summary, variants, extra):
```
Comment (ldoktor) [inline avocado/core/mux.py]: I actually did but then I worked on avocado-vt and said myself that English version is `extra`. So the question is whether the plugin writers will be more human or python. I have no direct preference here so if you prefer that, I'll change it.

Also about the `**` vs `argument` I thought it'd be clearer to hand over a custom dictionary with values rather than using kwargs. It'd better separate those two, but again, no preference here (the first version was actually like that)

File: avocado/core/job.py [hunk 17]
```diff
@@ -371,31 +370,15 @@ def _log_avocado_datadir():
         job_log.info('logs     ' + data_dir.get_logs_dir())
         job_log.info('')
 
-    def _log_variants_tree(self, variant):
```
Comment (clebergnu) [inline avocado/core/job.py]: That's my point.  In this specific commit, we loose it.  I doubt someone will checkout this commit and complain about it, but this indicates that the approach I suggested ealier (bring the new `str*()` utilities first, then replace the old functions with it) causes less (ideally no) disruption.

File: avocado/plugins/multiplex.py (hunk 11 above)
Comment (clebergnu) [inline avocado/plugins/multiplex.py]: I think we need to differentiate between 2 things:

1) Have the output or not (at a standard verbosity level)
2) The level of verbosity (which indeed has to be flexible to accommodate custom plugins)

So, if we expect the variants plugins to provide two *different* kinds of information, such as:

1) variants listing
2) variants summary

We should have options such as:

1) `--variants-list`
2) `--variants-summary`

When given, they will print output at a standard verbosity level.  To change that, other parameters options could be given:

1) `--variants-list=info`
2) `--variants-summary=debug`

If I were to code this, I'd align `info`, `debug`, etc with `logging.INFO` (numeric decimal 10), `logging.DEBUG` (numeric decimal 20), which would probably not surprise many people, and still allow custom plugins to define meanings for things such as value 15, 19, 25, etc.

File: avocado/core/parser.py (hunk 8 above)
Comment (clebergnu) [inline avocado/core/parser.py]: OK!

File: avocado/core/mux.py (hunk 6 above)
Comment (clebergnu) [inline avocado/core/mux.py]: Yes, I agree the change is atomic, but the documentation of the reason of the changes is a very very important thing.  And there are two ways to tackle that:  write a short commit message in a separate commit with the referred change (because it's then obvious), or give more context about the change in a larger commit, which I find more difficult to both authors and reviewers.

So, to summarize the discussion with stuff we already know:

1) describing the reason of the changes (if not obvious) is a *requirement* for commit messages
2) there's no "perfect size" when bundling changes and commits, so the use of one's *best* judgement when it comes to how to bundle changes into commits is *pretty important*

File: avocado/core/job.py (hunk 17 above)
Comment (ldoktor) [inline avocado/core/job.py]: Do you remember this: https://github.com/avocado-framework/avocado/pull/1774#discussion_r100579547 compare to this PR it was quite simple and still it was hard to just squash two patches. Now if I'm to not just squash two commits, but basically merge 2 renames, where the first one happened at the beginning and the second one after complete move of the files to a different location, that basically means I have to rework everything. And I'm not talking about minor clashes, that in deed means developing it all over again as all the code will be incompatible, even though the result will be the same.

This is why I was asking whether it is really worth requiring this change. It will most probably be never seen by people as it'll be atomically merged and if we want to revert a part, we'd most probably revert everything. So do you see this change so important that it means compete rework of all the patches which are here, or do you think it's low impact enough to allow this namechange to be kept? (I'm not arguing about some of the other issues, but the namechange which brakes everything is something I'd like to avoid).

File: avocado/plugins/multiplex.py (hunk 11 above)
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: well since the use of warn, error and critical would be probably not applicable here, this gives us the range of 20 with just 2 human friendly names. Tot his I'd prefer custom enum defined as "brief,normal,verbose,full". Still the numbers make more sense to me as we are dealing with different plugins and even our plugin actually accepts bitmask nowadays...

File: avocado/core/mux.py (hunk 6 above)
Comment (ldoktor) [inline avocado/core/mux.py]: Not sure I understand the `short commit` approach. If that means to extract this change and make it before moving the mux here than it'd be quite hard (not impossible, though). If it's just about saying this part is about to be removed because... than I don't like the idea.

Anyway would the extra note explaining the minor change of the inner API in the commit message be enough?

File: avocado/core/plugin_interfaces.py (hunk 1 above)
Comment (ldoktor) [inline avocado/core/plugin_interfaces.py]: Sure, as I mentioned earlier the later commits were added because the original code was not reviewed before I finished this commit. I'm fine with squashing those...

File: avocado/core/plugin_interfaces.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/plugin_interfaces.py]: yep

**Comment by ldoktor** [comment]
Thank you, @clebergnu, for the updates, I responded with 2 more questions and then I'm ready to rework this.

File: avocado/core/job.py (hunk 17 above)
Comment (clebergnu) [inline avocado/core/job.py]: It was not a matter of difficulty of squashing commits.  Given the changes that were requested (which were pretty minor), it was a matter of changing the commits in a way that the review of the next version would be more difficult.

Again, the changes in a commit have to stand for themselves in the commit itself, and *also* in the context of the PR.

File: avocado/plugins/multiplex.py (hunk 11 above)
Comment (clebergnu) [inline avocado/plugins/multiplex.py]: An interface (either a custom enum or reusing numbers for logging.*) as a guideline to plugin writers would be really helpful.  That will ensure some level of consistency among plugins.

File: avocado/core/mux.py (hunk 6 above)
Comment (clebergnu) [inline avocado/core/mux.py]: Sure, that falls into category 1 (describing the reason of the changes if not obvious).

**Comment by clebergnu** [comment]
Hey @ldoktor I think I responded to your pending questions.  Let me know if you need more info.

File: avocado/core/job.py (hunk 17 above)
Comment (ldoktor) [inline avocado/core/job.py]: Sure, I was just asking about the balance between nice and long-to-deliver and we already solved this on meeting. No doubts about this anymore 😄

File: avocado/plugins/multiplex.py (hunk 11 above)
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: OK, I'll provide a basic enum.

File: avocado/core/mux.py (hunk 6 above)
Comment (ldoktor) [inline avocado/core/mux.py]: Thanks, that helps 😄

**Comment by ldoktor** [comment]
Thank you, now the hard task in getting things right... (hopefully it won't end up like the Test cleanup where my minor changes produced thousand of little issues in every commit because of rerere...)

File: avocado/core/plugin_interfaces.py (hunk 1 above)
Comment (ldoktor) [inline avocado/core/plugin_interfaces.py]: done

File: avocado/core/plugin_interfaces.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/plugin_interfaces.py]: done

File: avocado/core/mux.py (hunk 15 above)
Comment (ldoktor) [inline avocado/core/mux.py]: done

File: avocado/core/varianter.py (hunk 14 above)
Comment (ldoktor) [inline avocado/core/varianter.py]: done

File: avocado/core/job.py [hunk 18]
```diff
@@ -371,7 +371,7 @@ def _log_avocado_datadir():
         job_log.info('')
 
     def _log_variants(self, variants):
-        lines = variants.str_variants(summary=True, contents=False)
+        lines = variants.str_variants(summary=1, variants=1)
```
Comment (ldoktor) [inline avocado/core/job.py]: done

File: avocado/core/mux.py (hunk 12 above)
Comment (ldoktor) [inline avocado/core/mux.py]: done

File: avocado/plugins/multiplex.py (hunk 11 above)
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: done

File: avocado/core/job.py (hunk 13 above)
Comment (ldoktor) [inline avocado/core/job.py]: done

File: avocado/core/mux.py (hunk 16 above)
Comment (ldoktor) [inline avocado/core/mux.py]: done

File: avocado/core/mux.py (hunk 10 above)
Comment (ldoktor) [inline avocado/core/mux.py]: done

File: avocado/core/mux.py (hunk 6 above)
Comment (ldoktor) [inline avocado/core/mux.py]: done

File: avocado/core/mux.py (hunk 7 above)
Comment (ldoktor) [inline avocado/core/mux.py]: merged with stevedore and the API is defined there.

File: avocado/core/parser.py (hunk 8 above)
Comment (ldoktor) [inline avocado/core/parser.py]: done

File: avocado/core/varianter.py (hunk 9 above)
Comment (ldoktor) [inline avocado/core/varianter.py]: done (squashed with stevedore)

File: avocado/plugins/multiplex.py (hunk 5 above)
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: done

File: avocado/core/job.py (hunk 17 above)
Comment (ldoktor) [inline avocado/core/job.py]: done
